mimxrt: Add the pin.irq() method to machine.pin.

robert-hh opened the pull request:
The API complies to the documentation. Both hard and soft irq are supported. Response times for hard irq are <1.5 µs for Teensy40/41 and <3 µs for MIMXRT1020-EVK. Response times for soft irq are <2.6 - 12 µs for Teensy 40/41 and <3.5-38µs for MIMXRT1010-EVK.
Due to the many GPIO ports, the RAM usage is rather high on advantage of response time.

robert-hh committed:
mimxrt: Support for pin.irq().

First flush: Compiles w/ errors, Tested for GPIO1
Next steps:
- Further tests
- Optimize RAM usage

robert-hh committed:
mimxrt: Pin IRQ optimizations.

- Reduce the size of the pin_irq_objects array. GPIO5 has three entries
  only, IO0 - IO2.
- Disable IRQ when called with handler=None

robert-hh committed:
mimxrt: Pin IRQ RAM size reduction.

Inherit board specific ROOT POINTERs allowing a smaller pin_irq_object
table for MIMXRT1010 and MIMXRT1020. The code stays as it is.

robert-hh committed:
mimxrt: Disable all pin IRQs on soft reset.

robert-hh committed:
mimxrt: More parameter checking for pin.irq().

And a tiny optimization in response time.

robert-hh committed:
mimxrt: Fix formatting complaint.

dpgeorge commented on ports/mimxrt/machine_pin.c:
```diff
@@ -234,6 +324,69 @@ STATIC mp_obj_t machine_pin_init(size_t n_args, const mp_obj_t *args, mp_map_t *
+    uint32_t gpio_nr = GPIO_get_instance(self->gpio);
+    uint32_t index = GET_PIN_IRQ_INDEX(gpio_nr, self->pin);
+    if (index >= ARRAY_SIZE(MP_STATE_PORT(machine_pin_irq_objects))) {
+        mp_raise_ValueError(MP_ERROR_TEXT("IQR not supported at that PIN"));
```
"IRQ not supported on given Pin"

robert-hh replied:
OK. Sounds better.

dpgeorge commented on ports/mimxrt/boards/MIMXRT1010_EVK/mpconfigboard.h:
```diff
@@ -8,3 +8,6 @@
 #define BOARD_FLASH_CONFIG_HEADER_H "evkmimxrt1010_flexspi_nor_config.h"
+
+#define MICROPY_BOARD_ROOT_POINTERS \
+    void *machine_pin_irq_objects[2 * 32];
```
would be simpler for a board to define the number of objects, eg `#define MICROPY_HW_NUM_PIN_IRQS (2 * 32)`, and then move this definition into `MICROPY_PORT_ROOT_POINTERS` in `mpconfigport.h`

robert-hh replied:
Yes, indeed.

dpgeorge commented on ports/mimxrt/machine_pin.c:
```diff
@@ -68,6 +69,95 @@ const mp_obj_type_t machine_pin_board_pins_obj_type = {
 
+STATIC const mp_irq_methods_t machine_pin_irq_methods;
+
+// STATIC const uint16_t GPIO_irqs[] = GPIO_IRQS;
```
remove commented out code

robert-hh replied:
OK

dpgeorge commented on ports/mimxrt/machine_pin.c:
```diff
@@ -234,6 +324,69 @@ STATIC mp_obj_t machine_pin_init(size_t n_args, const mp_obj_t *args, mp_map_t *
+        irq->base.ishard = args[ARG_hard].u_bool;
+        irq->flags = 0;
+        if (args[ARG_trigger].u_int >= ARRAY_SIZE(IRQ_mapping)) {
+            mp_raise_ValueError(MP_ERROR_TEXT("IQR mode not supported"));
```
IQR->IRQ

robert-hh replied:
That's one one my favorite permutations.

dpgeorge commented on ports/mimxrt/pin.h:
```diff
@@ -87,7 +88,14 @@ enum {
+    PIN_IRQ_RISING = 3,
+    PIN_IRQ_FALLING = 4,
+    PIN_IRQ_RISING_OR_FALLING = 0,
+};
```
looks like this enum can be removed

robert-hh replied:
I have to look into it, but most probably you are right.

robert-hh commented:
I'll make the changes, test them and supply later today (or your night). Yesterday evening I stumbled over time.time() missing. So I will complete the time module then. All pieces needed  for that are present.

robert-hh committed:
mimxrt: Implement changes according to the review.

- Move root pointer arary back to mpconfigport.h
- Fix typos
- Remove obsolete code

dpgeorge commented:
Squashed and merged in 7417c6ac91e47798b9c9eba2e7cd8f3dc1bbcc57 with minor edits, including a fix for the default irq trigger (was `kGPIO_IntRisingOrFallingEdge`, changed to 3).
